cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+  }
+
+  __syncthreads();
+
+  if ((threadIdx.x < num_point) && in_polygon) {
+    uint16_t num         = sums[threadIdx.x / num_threads_per_warp];
+    uint16_t warp_offset = __popc(vote >> (threadIdx.x % num_threads_per_warp)) - 1;
+    uint32_t pos         = mem_offset + num + warp_offset;
+
+    out_poly_idx.element<uint32_t>(pos)  = poly_idx;
+    out_point_idx.element<uint32_t>(pos) = tid;
+  }
+}
+
+template <typename T>
+std::unique_ptr<cudf::table> compute_quadtree_point_in_polygon(
+  cudf::column_view const &poly_indices,
+  cudf::column_view const &quad_indices,
+  cudf::column_view const &quad_lengths,
+  cudf::column_view const &quad_offsets,
+  cudf::column_view const &point_indices,
+  cudf::column_view const &point_x,
+  cudf::column_view const &point_y,
+  cudf::column_view const &poly_offsets,
+  cudf::column_view const &ring_offsets,
+  cudf::column_view const &poly_points_x,
+  cudf::column_view const &poly_points_y,
+  rmm::mr::device_memory_resource *mr,
+  cudaStream_t stream)
+{
+  auto num_original_pairs = poly_indices.size();
+  auto d_quad_lengths     = cudf::column_device_view::create(quad_lengths, stream);
+  auto get_num_units      = [quad_lengths = *d_quad_lengths] __device__(uint32_t quad_index) {
+    return ((quad_lengths.element<uint32_t>(quad_index) - 1) / threads_per_block) + 1;
+  };
+
+  // compute the total number of sub-pairs (units) using transform_reduce
+  uint32_t num_unit_pairs = thrust::transform_reduce(rmm::exec_policy(stream)->on(stream),
+                                                     quad_indices.begin<uint32_t>(),
+                                                     quad_indices.end<uint32_t>(),
+                                                     get_num_units,
+                                                     0,
+                                                     thrust::plus<uint32_t>());
+
+  auto d_quad_offset = [&]() {
+    // allocate memory for both numbers and their prefix-sums
+    rmm::device_uvector<uint32_t> d_num_sums(num_original_pairs, stream);
+    rmm::device_uvector<uint32_t> d_num_units(num_original_pairs, stream);
+
+    // computes numbers of sub-pairs for each quadrant-polygon pair
+    thrust::transform(rmm::exec_policy(stream)->on(stream),
+                      quad_indices.begin<uint32_t>(),
+                      quad_indices.end<uint32_t>(),
+                      d_num_units.begin(),
+                      get_num_units);
+
+    thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           d_num_units.begin(),
+                           d_num_units.end(),
+                           d_num_sums.begin());
+
+    // allocate memory for sub-pairs with four components:
+    // (polygon_idx, quadrant_idx, offset, length)
+
+    rmm::device_uvector<uint32_t> d_quad_offset(num_unit_pairs, stream);
+
+    thrust::fill(
+      rmm::exec_policy(stream)->on(stream), d_quad_offset.begin(), d_quad_offset.end(), 0);
```

Comment:
I think when zero-filling it's better to use `cudaMemsetAsync`, which I doubt Thrust uses.  (Faster in my tests)
 
```suggestion
    cudaMemsetAsync(d_quad_offset.data(), 0, num_unit_pairs * sizeof(uint32_t), stream);
```